src/BLSRegistryCoordinatorWithIndices.sol

```diff
@@ -337,109 +332,183 @@ contract BLSRegistryCoordinatorWithIndices is EIP712, Initializable, IBLSRegistr
+        require(!quorumsToRemove.isEmpty(), "BLSRegistryCoordinatorWithIndices._deregisterOperator: bitmap cannot be 0");
+        require(quorumsToRemove.isSubsetOf(currentBitmap), "BLSRegistryCoordinatorWithIndices._deregisterOperator: operator is not registered for specified quorums");
+        uint192 newBitmap = uint192(currentBitmap.minus(quorumsToRemove));
 
-        // Check if the operator is completely deregistering
-        bool completeDeregistration = previousBitmap == quorumsToRemoveBitmap;
+        /**
+         * Update operator's bitmap and status:
+         */
+        _updateOperatorBitmap({
+            operatorId: operatorId,
+            newBitmap: newBitmap
+        });
 
-        // Deregister operator with each of the registry contracts:
-        blsPubkeyRegistry.deregisterOperator(operator, quorumNumbersToRemove);
-        stakeRegistry.deregisterOperator(operatorId, quorumNumbersToRemove);
-        indexRegistry.deregisterOperator(operatorId, quorumNumbersToRemove);
-        
-        // If the operator still has active quorums, push a bitmap update.
-        // Otherwise, set them to deregistered
-        // TODO - change this to update history regardless
-        if (!completeDeregistration) {
-            _operatorBitmapHistory[operatorId].push(QuorumBitmapUpdate({
-                updateBlockNumber: uint32(block.number),
-                nextUpdateBlockNumber: 0,
-                quorumBitmap: previousBitmap & ~uint192(quorumsToRemoveBitmap) // this removes the quorumsToRemoveBitmap from the quorumBitmapBeforeUpdate
-            }));
-        } else {
+        // If the operator is no longer registered for any quorums, update their status
+        if (newBitmap.isEmpty()) {
             operatorInfo.status = OperatorStatus.DEREGISTERED;
             emit OperatorDeregistered(operator, operatorId);
         }
+
+        // Deregister operator with each of the registry contracts:
+        blsPubkeyRegistry.deregisterOperator(operator, quorumNumbers);
+        stakeRegistry.deregisterOperator(operatorId, quorumNumbers);
+        indexRegistry.deregisterOperator(operatorId, quorumNumbers);
+    }
+
+    /**
+     * @notice Returns the stake threshold required for an incoming operator to replace an existing operator
+     * The incoming operator must have more stake than the return value.
+     */
+    function _individualKickThreshold(uint96 operatorStake, OperatorSetParam memory setParams) internal pure returns (uint96) {
+        return operatorStake * setParams.kickBIPsOfOperatorStake / BIPS_DENOMINATOR;
+    }
+
+    /**
+     * @notice Returns the total stake threshold required for an operator to remain in a quorum.
+     * The operator must have at least the returned stake amount to keep their position.
+     */
+    function _totalKickThreshold(uint96 totalStake, OperatorSetParam memory setParams) internal pure returns (uint96) {
+        return totalStake * setParams.kickBIPsOfTotalStake / BIPS_DENOMINATOR;
```

Comment:
might want public getters that use these + read from state?

Reply:
I'm open to it if it's needed.